crates/sui-source-validation/src/lib.rs

```diff
@@ -0,0 +1,238 @@
+    ) -> Result<(), DependencyVerificationError> {
+        let compiled_dep_map = get_module_bytes_map(&compiled_package);
+
+        let mut on_chain_module_count = 0usize;
+        let mut verified_dependencies: HashMap<AccountAddress, Dependency> = HashMap::new();
+
+        for (pkg_symbol, (addr, local_pkg_bytes)) in compiled_dep_map {
+            if pkg_symbol == compiled_package.compiled_package_info.package_name {
+                continue;
+            };
+
+            // fetch the Sui object at the address specified for the package in the local resolution table
+            let on_chain_package = self.pkg_for_address(&addr).await?;
+
+            for (module_symbol, local_bytes) in local_pkg_bytes {
+                // package addresses may show up many times, but we only need to verify them once
+                // zero address is the package we're checking dependencies for
+                if verified_dependencies.contains_key(&addr) || addr == AccountAddress::ZERO {
+                    continue;
+                }
+
+                let mod_str = module_symbol.to_string();
+                let on_chain_bytes = match on_chain_package.module_map.get(&mod_str) {
+                    Some(oc_bytes) => oc_bytes,
+                    None => {
+                        return Err(DependencyVerificationError::OnChainDependencyNotFound(
+                            pkg_symbol,
+                            Some(module_symbol),
+                        ))
+                    }
+                };
+
+                // compare local bytecode to on-chain bytecode to ensure integrity of our dependencies
+                if &local_bytes != on_chain_bytes {
+                    return Err(DependencyVerificationError::ModuleBytecodeMismatch(
+                        pkg_symbol.to_string(),
+                        module_symbol.to_string(),
+                        addr,
+                    ));
+                }
+
+                on_chain_module_count += 1;
+
+                if self.verbose {
+                    println!(
+                        "{}::{} - {} bytes, code matches",
+                        pkg_symbol,
+                        module_symbol,
+                        on_chain_bytes.len()
+                    );
+                }
+            }
+
+            verified_dependencies.insert(
+                addr,
+                Dependency {
+                    symbol: pkg_symbol.to_string(),
+                    module_bytes: on_chain_package.module_map.clone(),
+                },
+            );
+        }
+
+        // total number of modules in packages must match, in addition to each individual module matching
+        let len = compiled_package.deps_compiled_units.len();
+        // only need to check for less than, because if local modules are missing on-chain we've
+        // already errored out with a OnChainDependencyNotFound error
+        if len < on_chain_module_count {
```

Comment:
+1 here with @amnn comment about the Set-based check below. Turning an `O(n)` check to an `O(1)` would be nice :) I would probably just build a set of all expected dependencies, remove them as you go, and then print them out here if the set is not empty.